### PR TITLE
Turn off real-time thread throttling.

### PR DESCRIPTION
**Pull request by vext01**
```
Subject: [krun:bencher3] dmesg changed                                                              
                                                                                                    
Message from krun running on bencher3:                                                              
                                                                                                    
dmesg seems to have changed! Diff follows:                                                          
--- 2016-04-14 23:52:42                                                                             
+++ 2016-04-15 00:32:48                                                                             
@@ -830,3 +830,4 @@                                                                                 
 [    3.830275] [drm] Enabling RC6 states: RC6 on, RC6p off, RC6pp off                              
 [   15.530039] r8169 0000:06:00.0 eth0: link up                                                    
 [   15.530044] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready                             
+[  157.924835] [sched_delayed] sched: RT throttling activated                                      
```

Translation: that real-time process is *stuck* on the CPU and no other process can get a word in. I'm going to start artificially pre-empting it so other processes have a chance.

http://man7.org/linux/man-pages/man7/sched.7.html

This change disables this behaviour. We want the benchmark to be as uninterrupted as possible.

I've not tested this on the warmup experiment for real. I suggest (if this looks good) we merge this and back it out if we have any problems later.

### CODE REVIEW

File: krun/platform.py [hunk 1]
```diff
@@ -647,6 +648,49 @@ def check_preliminaries(self):
...
+
+                cmd = " ".join(args)
+                run_shell_cmd(cmd)
+                again = True
+            else:
+                fatal("Could not set %s to -1" %
+                      LinuxPlatform.SCHED_RT_RUNTIME_US)
+
+        # We changed something. Go around again and check it took effect.
+        if again:
```
Comment (ltratt) [inline krun/platform.py]: A recursive call for something that should be a loop is dangerous (and harder to understand).

File: krun/platform.py [hunk 2]
```diff
@@ -647,6 +648,49 @@ def check_preliminaries(self):
...
+        self._check_realtime_throttle_disabled()
+
+    def _check_realtime_throttle_disabled(self, first=True):
+        """Linux kernel gets pretty upset if you run a CPU intensive thread
+        under the real-time thread schedule policy. By default Linux will
+        artificially pre-empt such threads to give other things a chance to run
+        on this core. A switch will flip at runtime leaving a message in dmesg
+        when this comes into effect.
+
+        See "Limiting the CPU usage of real-time and deadline processes" here:
```
Comment (snim2) [inline krun/platform.py]: was there meant to be a URL after `here:`?

File: krun/platform.py [hunk 3]
```diff
@@ -647,6 +648,49 @@ def check_preliminaries(self):
         self._check_perf_samplerate()
         self._check_tickless_kernel()
         self._check_aslr_disabled()
+        self._check_realtime_throttle_disabled()
+
+    def _check_realtime_throttle_disabled(self, first=True):
```
Comment (snim2) [inline krun/platform.py]: I didn't guess what `first` was for until I got to the end of this method. Can we call it something a bit more verbose, like `first_attempt` or `first_check`?

**Comment by snim2** [comment]
The code here looks good. A couple of very minor points, but otherwise this looks ready to go.

**Comment by vext01** [comment]
Aside from code, does the approach sound reasonable?

**Comment by snim2** [comment]
I think so, but we won't really know until we can try it, I think.

**Comment by ltratt** [comment]
I guess we won't know until we try it.

**Comment by vext01** [comment]
OK to squash?

**Comment by snim2** [comment]
Squash away!

**Comment by vext01** [comment]
squelch.

**Comment by vext01** [comment]
Wait typo

**Comment by vext01** [comment]
Fixed, go ahead.

**Comment by vext01** [comment]
Cheers

**Pull request merged**